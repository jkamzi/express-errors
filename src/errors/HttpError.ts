export default class HttpError extends Error {
  public status: number;

  public reason: string;

  constructor(status: number, message: string, reason?: string) {
    super(message);
    this.status = status;
    this.reason = reason || message;

    this.name = 'HttpError';
  }
}
