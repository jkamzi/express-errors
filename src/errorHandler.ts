import { Request, Response, ErrorRequestHandler } from 'express';
import HttpError from './errors/HttpError';
import { NextFunction } from 'connect';

type ErrorType = Error | HttpError;

export type Logger = (err: HttpError) => Promise<void> | void;

export type Options = Readonly<{
  logger?: undefined | Logger;
}>;

export type ErrorResponse = Readonly<{
  error: {
    message: string;
    reason: string;
    status: number;
  };
}>;

function isHttpError(err: ErrorType): err is HttpError {
  if (err instanceof HttpError) {
    return true;
  }

  return false;
}

export function errorBody(err: HttpError): ErrorResponse {
  return {
    error: {
      message: err.message,
      reason: err.reason,
      status: err.status,
    },
  };
}

export default function errorHandler(options?: Options): ErrorRequestHandler {
  const defaultOptions: Options = {
    logger: undefined,
    ...options,
  };
  return async (
    err: Error | HttpError,
    req: Request,
    res: Response,
    next: NextFunction,
  ) => {
    if (isHttpError(err)) {
      if (defaultOptions.logger) {
        await defaultOptions.logger(err);
      }
      return res.status(err.status).json(errorBody(err));
    }

    return next(err);
  };
}
