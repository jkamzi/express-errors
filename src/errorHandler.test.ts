import { Request, Response } from 'express';
import type { Logger } from './types';
import errorHandler from './errorHandler';
import HttpError from './errors/HttpError';

describe('errorHandler', () => {
  it('should call next() when error is not HttpError', () => {
    const next = jest.fn();
    const err = new Error('Error');

    errorHandler()(err, {} as Request, {} as Response, next);

    expect(next).toHaveBeenCalledWith(err);
  });

  it('should be possible to replace error output format', () => {
    const next = jest.fn();
    const res: Partial<Response> = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn(),
    };
    const err = new HttpError(400, 'Bad Request', 'Who knows');

    errorHandler({
      formatter: (err: HttpError) => ({
        error: {
          message: err.message,
          stack: err.stack,
        },
      }),
    })(err, {} as Request, res as Response, next);
    expect(res.json).toHaveBeenCalledWith({
      error: {
        message: err.message,
        stack: err.stack,
      },
    });
  });

  it('should return response when error is HttpError', () => {
    const next = jest.fn();
    const res: Partial<Response> = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn(),
    };
    const err = new HttpError(400, 'Bad Request', 'Who knows');

    errorHandler()(err, {} as Request, res as Response, next);

    expect(res.status).toHaveBeenCalledWith(400);
    expect(res.json).toHaveBeenCalledWith({
      error: {
        message: err.message,
        reason: err.reason,
        status: err.status,
      },
    });
    expect(next).not.toHaveBeenCalled();
  });

  /**
   * Not sure if this is the best way to test this.
   */
  describe('logger', () => {
    it('should accept async logger', () => {
      const next = jest.fn();
      const logger = jest.fn();
      const res: Partial<Response> = {
        status: jest.fn().mockReturnThis(),
        json: jest.fn(),
      };

      const err = new HttpError(400, 'Bad Request', 'Who knows');
      const asyncLogger: Logger = async (err: HttpError): Promise<void> => {
        return new Promise((resolve) => {
          logger();
          return resolve();
        });
      };

      errorHandler({
        logger: asyncLogger,
      })(err, {} as Request, res as Response, next);

      expect(logger).toHaveBeenCalled();
    });

    it('should accept sync logger', () => {
      const next = jest.fn();
      const logger = jest.fn();
      const res: Partial<Response> = {
        status: jest.fn().mockReturnThis(),
        json: jest.fn(),
      };

      const err = new HttpError(400, 'Bad Request', 'Who knows');
      const syncLogger: Logger = (err: HttpError): void => {
        logger();
      };

      errorHandler({
        logger: syncLogger,
      })(err, {} as Request, res as Response, next);

      expect(logger).toHaveBeenCalled();
    });
  });
});
