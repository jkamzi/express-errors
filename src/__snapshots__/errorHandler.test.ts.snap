// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorHandler errorBody() returns error object 1`] = `
Object {
  "error": Object {
    "message": "Bad Request",
    "reason": "Who knows",
    "status": 400,
  },
}
`;
